lib/Dialect/Arith/Conversions/ArithToCGGI/ArithToCGGI.cpp

```diff
@@ -0,0 +1,232 @@
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mlir::arith::ExtUIOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto outType = convertArithToCGGIType(
+        cast<IntegerType>(op.getResult().getType()), op->getContext());
+    auto castOp = b.create<cggi::CastOp>(op.getLoc(), outType, adaptor.getIn());
+
+    rewriter.replaceOp(op, castOp);
+    return success();
+  }
+};
+
+struct ConvertShRUIOp : public OpConversionPattern<mlir::arith::ShRUIOp> {
+  ConvertShRUIOp(mlir::MLIRContext *context)
+      : OpConversionPattern<mlir::arith::ShRUIOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mlir::arith::ShRUIOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto cteShiftSizeOp = op.getRhs().getDefiningOp<mlir::arith::ConstantOp>();
+
+    if (cteShiftSizeOp) {
+      auto outputType = adaptor.getLhs().getType();
+
+      auto shiftAmount = cast<IntegerAttr>(cteShiftSizeOp.getValue())
+                             .getValue()
+                             .getSExtValue();
+
+      auto inputValue =
+          mlir::IntegerAttr::get(rewriter.getI8Type(), (int8_t)shiftAmount);
+      auto cteOp = rewriter.create<mlir::arith::ConstantOp>(
+          op.getLoc(), rewriter.getI8Type(), inputValue);
+
+      auto shiftOp =
+          b.create<cggi::ShiftRightOp>(outputType, adaptor.getLhs(), cteOp);
+      rewriter.replaceOp(op, shiftOp);
+
+      return success();
+    }
+
+    cteShiftSizeOp = op.getLhs().getDefiningOp<mlir::arith::ConstantOp>();
+
+    auto outputType = adaptor.getRhs().getType();
+
+    auto shiftAmount =
+        cast<IntegerAttr>(cteShiftSizeOp.getValue()).getValue().getSExtValue();
+
+    auto inputValue = mlir::IntegerAttr::get(rewriter.getI8Type(), shiftAmount);
+    auto cteOp = rewriter.create<mlir::arith::ConstantOp>(
+        op.getLoc(), rewriter.getI8Type(), inputValue);
+
+    auto shiftOp =
+        b.create<cggi::ShiftRightOp>(outputType, adaptor.getLhs(), cteOp);
+    rewriter.replaceOp(op, shiftOp);
+    rewriter.replaceOp(op.getLhs().getDefiningOp(), cteOp);
+
+    return success();
+  }
+};
+
+struct ArithToCGGI : public impl::ArithToCGGIBase<ArithToCGGI> {
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto *module = getOperation();
+    ArithToCGGITypeConverter typeConverter(context);
+
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    target.addLegalDialect<cggi::CGGIDialect>();
+    target.addIllegalDialect<mlir::arith::ArithDialect>();
+
+    target.addDynamicallyLegalOp<mlir::arith::ConstantOp>(
+        [](mlir::arith::ConstantOp op) {
+          // Allow use of constant if it is used to denote the size of a shift
```

Comment:
(optional): This is not too bad, but one way we could avoid it is by making the CGGI shift op take an attribute as its shift instead of a value. Is there any time when the shift amount is not statically known?